lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,607 @@
+
+    local spacing = (self._private.spacing or 0)
+        + (self._private.border_width or 0)
+
+    local y = 0
+
+    for row_idx = 1, #max_row do -- Vertical
+        local x = 0
+        for col_idx = 1, #max_col do -- Horizontal
+            local v = self._private.rows[row_idx][col_idx]
+
+            if v then
+                table.insert(result, base.place_widget_at(
+                    v, x, y, max_col[col_idx], max_row[row_idx]
+                ))
+                x = x + max_col[col_idx] + spacing
+            end
+        end
+
+        y = y + max_row[row_idx] + spacing
+    end
+
+    return result
+end
+
+function grid:after_draw_children(context, cr, width, height)
+    -- TODO support transparent colors
+    if self._private.border_width and self._private.border_width > 0 then
+        local bc = self._private.border_color
+
+        if bc then
+            cr:set_source(bc)
+        end
+
+        local bw = self._private.border_width
+        local max_col, max_row, mw, mh = get_ajusted_cell_sizes(
+            self, context, width, height
+        )
+        local spacing = (self._private.spacing or 0)
+            + (self._private.border_width or 0)
+
+        cr:set_line_width(bw)
+
+
+        local sw, sh = 0, 0
+
+        local max_x = math.min(width-bw , mw+#max_col*spacing)
+        local max_y = math.min(height-bw, mh+#max_col*spacing)
+
+        cr:rectangle( bw/2, bw/2, max_x, max_y )
+
+        for i=1, #max_row do
+            cr:move_to(bw/2, math.max(0, sh - spacing/2))
+            cr:line_to(max_x, math.max(0, sh - spacing/2))
+
+            sh = sh + max_row[i] + spacing
+        end
+
+        for i=1, #max_col do
+            cr:move_to(math.max(0, sw - spacing/2), bw/2)
+            cr:line_to(math.max(0, sw - spacing/2), max_y)
+
+            sw = sw + max_col[i] + spacing
+        end
+
+        cr:stroke()
+    end
+end
+
+local function get_next_empty(self, row, column)
+    row, column = row or 1, column or 1
+    local cc = self._private.column_count
+    for i = row, math.huge do
+        local r = self._private.rows[i]
+
+        if not r then
+            r = {}
+            self._private.rows[i] = r
+            self._private.row_count = i
+
+            return i, 1
+        end
+
+        for j = column, cc do
+            if not r[j] then
+                return i, j
+            end
+        end
+    end
+end
+
+local function index_to_point(data, index)
+    local counter = 0
+
+    -- As the widget matrix is sparse, the modulo cannot be used
+    for i=1, data._private.row_count do
+        for j=1, data._private.column_count do
+            counter = counter + (data._private.rows[i][j] and 1 or 0)
+
+            if counter == index then
+                return i, j
+            end
+        end
+    end
+
+    -- This can have bad side effects, but is the best fallback
+    return math.floor(index / data._private.row_count) + 1,
+        index % data._private.column_count + 1
```

Comment:
The only caller of this function is `:remove`. I have to suggestions:
- Get rid of this function as a separate function and instead inline it into `:remove`
- Handle this "index out of bounds" case via `return` and doing nothing instead of this bad code (every widget has exactly one index... unless you specify an invalid index that is too large, then some random other widget is hit)